src/metrics/prometheus.go

```diff
@@ -0,0 +1,207 @@
+// +build prometheus
+
+// Package metrics contains support for reporting metrics to an external server,
+// currently a Prometheus pushgateway. Because plz runs as a transient process
+// we can't wait around for Prometheus to call us, we've got to push to them.
+package metrics
+
+import (
+	"os/user"
+	"runtime"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/push"
+	"gopkg.in/op/go-logging.v1"
+
+	"core"
+)
+
+var log = logging.MustGetLogger("metrics")
+
+// This is the maximum number of errors after which plz will stop attempting to send metrics.
+const maxErrors = 3
+
+type metrics struct {
+	url                                           string
+	newMetrics                                    bool
+	stopChan                                      chan bool
+	cancelled                                     bool
+	errors                                        int
+	pushes                                        int
+	buildCounter, cacheCounter, testCounter       *prometheus.CounterVec
+	buildHistogram, cacheHistogram, testHistogram *prometheus.HistogramVec
+}
+
+// m is the singleton metrics instance.
+var m *metrics
+
+// InitFromConfig sets up the initial metrics from the configuration.
+func InitFromConfig(config *core.Configuration) {
+	if config.Metrics.PushGatewayURL != "" {
+		m = initMetrics(config.Metrics.PushGatewayURL, config.Metrics.PushFrequency)
+		prometheus.MustRegister(m.buildCounter)
+		prometheus.MustRegister(m.cacheCounter)
+		prometheus.MustRegister(m.testCounter)
+		prometheus.MustRegister(m.buildHistogram)
+		prometheus.MustRegister(m.cacheHistogram)
+		prometheus.MustRegister(m.testHistogram)
+	}
+}
+
+// initMetrics initialises a new metrics instance.
+// This is deliberately not exposed but is useful for testing.
+func initMetrics(url string, frequency int) *metrics {
+	u, err := user.Current()
+	if err != nil {
+		log.Warning("Can't determine current user name for metrics")
+		u = &user.User{Username: "unknown"}
+	}
+	constLabels := prometheus.Labels{
```

Comment:
> plz is agnostic to the version control system so at the moment we never know the branch. I suppose we could quickly run a cheeky git command to find it.
Hm, I thought this might be the case :/ I think a better solution is to have a section in the plzconfig about the additional dimensions/labels to apply to each metric, although that feels like a more substantial piece of work.

Re master/non-master -- definitely agree, that was my thinking, too.